Numerical accuracy

Compression tests fail randomly for different instances.

- Is the assert_mpa_identical too strict?
- We should check seeding since this is not reproducible.

Apparently we have another test which sometimes fails but sometimes does not:

https://travis-ci.org/dseuss/mpnum/jobs/126319313


```
=================================== FAILURES ===================================
_____________________ [doctest] mpnum.factory.random_mpdo ______________________
290     :returns: randomly choosen matrix product (pure) state
291 
292     >>> rho = random_mpdo(4, 2, 4)
293     >>> rho.bdims, rho.pdims
294     ((4, 4, 4), ((2, 2), (2, 2), (2, 2), (2, 2)))
295     >>> rho.normal_form
296     (0, 4)
297     >>> rho_dense = rho.to_array_global().reshape((2**4,) * 2)
298     >>> from numpy.linalg import eigvals
299     >>> min(eigvals(rho_dense)) >= -1e-14
Expected:
    True
Got:
    False
```

vs.

https://travis-ci.org/dseuss/mpnum/jobs/126319278 (1 unrelated commit earlier, all good)
https://travis-ci.org/dseuss/mpnum/jobs/126640751 (1 unrelated commit later, all good)

--- a/mpnum/factory.py
+++ b/mpnum/factory.py
@@ -287,17 +287,13 @@ def random_mpdo(sites, ldim, bdim, randstate=np.random):
     :param ldim: Local dimension
     :param bdim: Bond dimension
     :param randstate: numpy.random.RandomState instance
-    :returns: randomly choosen matrix product (pure) state
+    :returns: randomly choosen classicaly correlated matrix product density op.
 
     >>> rho = random_mpdo(4, 2, 4)
     >>> rho.bdims, rho.pdims
     ((4, 4, 4), ((2, 2), (2, 2), (2, 2), (2, 2)))
     >>> rho.normal_form
     (0, 4)
-    >>> rho_dense = rho.to_array_global().reshape((2**4,) * 2)
-    >>> from numpy.linalg import eigvals
-    >>> min(eigvals(rho_dense)) >= -1e-14
-    True
 
     """
     # generate density matrix as a mixture of `bdim` pure product states

--- a/setup.py
+++ b/setup.py
@@ -28,38 +28,37 @@ class PyTest(Command):
     description = "Runs test suite"
     user_options = [
         ('selector=', None, "Specifies the tests to run"),
-        ('covreport', None, "Run coverage report from tests")
+        ('covreport', None, "Run coverage report from tests"),
+        ('pdb', None, "Whether to run pdb on failure"),
     ]
 
     def initialize_options(self):
         self.selector = 'not long'
         self.covreport = None
+        self.pdb = False
 
     def finalize_options(self):
         pass
 
     def run(self):
         import pytest
-        # Disable numba JIT optimization, which will take a couple of seconds
-        # (i.e. as long as the tests are supposed to take).  This must be done
-        # before importing numba.
-        os.environ['NUMBA_DISABLE_JIT'] = '1'
+        args = [] if not self.pdb else ['--pdb']
 
         if self.covreport is None:
             # Run doctests afterwards because they auto-import things. We want
             # to see missing imports.
-            errno = pytest.main(['-m', self.selector, 'tests'])
+            errno = pytest.main(['-m', self.selector, 'tests'] + args)
             if errno != 0:
                 raise SystemExit(errno)
 
-            errno = pytest.main(['--doctest-modules', 'mpnum'])
+            errno = pytest.main(['--doctest-modules', 'mpnum'] + args)
             if errno != 0:
                 raise SystemExit(errno)
 
         else:
             errno = pytest.main(['-m', self.selector, '--cov-report', 'term',
                                  '--cov-report', 'html', '--cov=mpnum',
-                                 '--doctest-modules', 'mpnum', 'tests'])
+                                 '--doctest-modules', 'mpnum', 'tests'] + args)
             if errno != 0:
                 raise SystemExit(errno)
 

--- a/tests/factory_test.py
+++ b/tests/factory_test.py
@@ -1,0 +1,22 @@
+# encoding: utf-8
+
+
+from __future__ import division, print_function
+
+import numpy as np
+import pytest as pt
+
+import mpnum.factory as factory
+import numpy as np
+
+
+@pt.mark.parametrize('nr_sites, local_dim, bond_dim', [(2, 3, 3), (3, 2, 4),
+                                                       (6, 2, 4), (4, 3, 5),
+                                                       (5, 2, 1)])
+def test_mpdo_positivity(nr_sites, local_dim, bond_dim, rgen):
+    rho = factory.random_mpdo(nr_sites, local_dim, bond_dim, rgen)
+    rho_dense = rho.to_array_global().reshape((local_dim**nr_sites,) * 2)
+
+    np.testing.assert_array_almost_equal(rho_dense, rho_dense.conj().T)
+    lambda_min = min(np.real(np.linalg.eigvals(rho_dense)))
+    assert lambda_min > -1e-14, "{} < -1e-14".format(lambda_min)

